### PR TITLE
refactor: avoid memory allocation during partition write requests

### PR DESCRIPTION
**Pull request by ShiKaiWi**
## Rationale
Some memory allocation during partitioning write request can be avoided.

## Detailed Changes
Introduce an enum called `PartitionedRows` as the return type of `PartitionRule.locate_partitions_for_write` to avoid unnecessary memory allocation.

## Test Plan
Existing test.

### CODE REVIEW

File: partition_table_engine/src/partition.rs [hunk 1]
```diff
@@ -84,6 +99,86 @@ impl PartitionTableImpl {
...
+        schema: Schema,
+        partitioned_rows: PartitionedRowsIter,
+    ) -> Result<usize> {
+        let mut split_rows = HashMap::new();
+        for PartitionedRow { partition_id, row } in partitioned_rows {
+            split_rows
+                .entry(partition_id)
+                .or_insert_with(Vec::new)
+                .push(row);
+        }
```
Comment (Rachelint) [inline partition_table_engine/src/partition.rs:135]: How about move here into `location_partitions_for_write`, and just return final `HashMap<partition, RowGroup>`?

File: partition_table_engine/src/partition.rs [hunk 2]
```diff
@@ -84,6 +99,86 @@ impl PartitionTableImpl {
...
+
+        // Insert split write request through remote engine.
+        let mut request_batch = Vec::with_capacity(split_rows.len());
+        for (partition, rows) in split_rows {
+            let sub_table_ident = self.get_sub_table_ident(partition);
+            let row_group = RowGroupBuilder::with_rows(schema.clone(), rows)
```
Comment (jiacai2050) [inline partition_table_engine/src/partition.rs:141]: With_rows invoke ` Self::new(schema);` inside, `with_capacity` is more suitable.

File: table_engine/src/partition/rule/mod.rs [hunk 3]
```diff
@@ -20,19 +20,38 @@ mod filter;
...
 
-use common_types::{datum::DatumKind, row::RowGroup};
+use common_types::row::{Row, RowGroup};
 
 use self::filter::PartitionFilter;
 use crate::partition::Result;
 
+/// The partitioned rows of the written requests.
+pub enum PartitionedRows {
+    One {
```
Comment (jiacai2050) [inline table_engine/src/partition/rule/mod.rs:30]: `Single` is more suitable than `One`?

File: table_engine/src/partition/rule/mod.rs [hunk 4]
```diff
@@ -48,17 +67,4 @@ pub trait PartitionRule: Send + Sync + 'static {
...
-}
-
-impl ColumnWithType {
-    pub fn new(column: String, datum_type: DatumKind) -> Self {
-        Self { column, datum_type }
-    }
-}
-
-pub type PartitionRuleRef = Box<dyn PartitionRule>;
+pub type PartitionRulePtr = Box<dyn PartitionRule>;
```
Comment (jiacai2050) [inline table_engine/src/partition/rule/mod.rs:70]: IMO `PartitionRuleRef` is better, in safe Rust we don't operate on pointers.

Comment (ShiKaiWi) [reply table_engine/src/partition/rule/mod.rs:70]: I guess `Ref` is usually used for `Arc`?

File: partition_table_engine/src/partition.rs (hunk 2 above)
Comment (ShiKaiWi) [reply partition_table_engine/src/partition.rs:141]: I guess `RowGroupBuilder` does nothing with this PR.

File: table_engine/src/partition/rule/mod.rs (hunk 3 above)
Comment (ShiKaiWi) [reply table_engine/src/partition/rule/mod.rs:30]: Done.

**Review by jiacai2050 — APPROVED** [review]
LGTM

**Pull request merged**